xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp

```diff
@@ -445,8 +444,273 @@ bool CVaapiTexture::TestInteropHevc(VADisplay vaDpy, EGLDisplay eglDisplay)
+
+  for (int layerNo = 0; layerNo < surface.num_layers; layerNo++)
+  {
+    int plane = 0;
+    auto const& layer = surface.layers[layerNo];
+    if (layer.num_planes != 1)
+    {
+      CLog::LogFunction(LOGDEBUG, "CVaapiTexture::MapEsh", "DRM-exported layer has %d planes - only 1 supported", layer.num_planes);
+      return false;
+    }
+    auto const& object = surface.objects[layer.object_index[plane]];
+
+    MappedTexture* texture{};
+    EGLint width{m_textureSize.Width()};
+    EGLint height{m_textureSize.Height()};
+
+    switch (surface.num_layers)
+    {
+      case 2:
+        switch (layerNo)
+        {
+          case 0:
+            texture = &m_y;
+            break;
+          case 1:
+            texture = &m_vu;
+            if (surface.fourcc == VA_FOURCC_NV12 || surface.fourcc == VA_FOURCC_P010 || surface.fourcc == VA_FOURCC_P016)
+            {
+              // Adjust w/h for 4:2:0 subsampling on UV plane
+              width = (width + 1) >> 1;
+              height = (height + 1) >> 1;
+            }
+            break;
+          default:
+            throw std::logic_error("Impossible layer number");
+        }
+        break;
+      default:
+        CLog::LogFunction(LOGDEBUG, "CVaapiTexture::MapEsh", "DRM-exported surface %d layers - only 2 supported", surface.num_layers);
+        return false;
+    }
+
+    CEGLAttributes<8> attribs; // 6 static + 2 modifiers
+    attribs.Add({
+      {EGL_LINUX_DRM_FOURCC_EXT, static_cast<EGLint> (layer.drm_format)},
+      {EGL_WIDTH, static_cast<EGLint> (width)},
+      {EGL_HEIGHT, static_cast<EGLint> (height)},
+      {EGL_DMA_BUF_PLANE0_FD_EXT, object.fd},
+      {EGL_DMA_BUF_PLANE0_OFFSET_EXT, static_cast<EGLint> (layer.offset[plane])},
+      {EGL_DMA_BUF_PLANE0_PITCH_EXT, static_cast<EGLint> (layer.pitch[plane])}
+    });
+
+    if (m_hasPlaneModifiers)
+    {
+      attribs.Add({
+        {EGL_DMA_BUF_PLANE0_MODIFIER_LO_EXT, static_cast<EGLint> (object.drm_format_modifier)},
+        {EGL_DMA_BUF_PLANE0_MODIFIER_HI_EXT, static_cast<EGLint> (object.drm_format_modifier >> 32)}
+      });
+    }
+
+    texture->eglImage = m_interop.eglCreateImageKHR(m_interop.eglDisplay,
+                                                   EGL_NO_CONTEXT, EGL_LINUX_DMA_BUF_EXT, nullptr,
+                                                   attribs.Get());
+    if (!texture->eglImage)
+    {
+      CEGLUtils::LogError("Failed to import VA DRM surface into EGL image");
+      return false;
+    }
+
+    texture->glTexture = ImportImageToTexture(texture->eglImage);
+  }
+
+  return true;
+#else
+  return false;
+#endif
+}
+
+void CVaapi2Texture::Unmap(MappedTexture& texture)
+{
+  if (texture.eglImage != EGL_NO_IMAGE_KHR)
+  {
+    m_interop.eglDestroyImageKHR(m_interop.eglDisplay, texture.eglImage);
+    texture.eglImage = EGL_NO_IMAGE_KHR;
+    glDeleteTextures(1, &texture.glTexture);
+  }
+}
+
+void CVaapi2Texture::Unmap()
+{
+  if (!m_vaapiPic)
+    return;
+
+  Unmap(m_y);
```

Comment:
I can rename it if you find it misleading. The benefit of having it is not having to repeat code.

Reply:
the code is easier to read if the function is expanded here. this I rank higher than saving 2 lines of code.
the desire of coders to compact code in early stages is one reason for this spaghetti code we've got. next step is often to create a parameter with a default value when reuse of a method does not exactly fit.
you won't find a suitable name because your primary motivation is grouping random parts of code for reuse.

Reply:
> the desire of coders to compact code in early stages is one reason for this spaghetti code we've got.

I don't see how this is spaghetti code, and I'm not sure whether you are suggesting it is or just fear that it will turn into such later?

> you won't find a suitable name because your primary motivation is grouping random parts of code for reuse.

It unmaps a texture that was mapped into EGL+GL from DRM. It's a decomposition of the `CVaapiTexture::Unmap` function (that unmaps two textures) into one unmap step for each of the two textures. If `Unmap` is not a fitting name, then `CVaapiTexture::Unmap` is already unfitting.

In principle, I get your argument. Grouping "random parts of code" into functions isn't helpful. However, I don't see why you think it is random (i.e. it does not have one clearly defined purpose) here.

Reply:
> I don't see how this is spaghetti code, and I'm not sure whether you are suggesting it is or just fear that it will turn into such later?

I was referring to kodi code in general and what I have seen during the last couple of years. 
Since you wrote this code, you know what "unmap" does. For somebody who may study this code in the future it means that she/he must navigate to this function in order to understand what it does.

Reply:
btw: don't forget to drop those methods